pynsot/commands/cmd_protocol_types.py

```diff
@@ -0,0 +1,340 @@
+# -*- coding: utf-8 -*-
+
+"""
+Sub-command for Protocol Types.
+
+In all cases ``data = ctx.params`` when calling the appropriate action method
+on ``ctx.obj``. (e.g. ``ctx.obj.add(ctx.params)``)
+
+Also, ``action = ctx.info_name`` *might* reliably contain the name of the
+action function, but still not sure about that. If so, every function could be
+fundamentally simplified to this::
+
+    getattr(ctx.obj, ctx.info_name)(ctx.params)
+"""
+
+from __future__ import unicode_literals
+import logging
+
+from ..vendor import click
+from . import callbacks, types
+from .cmd_protocols import DISPLAY_FIELDS as PROTOTYPE_DISPLAY_FIELDS
+
+
+# Logger
+log = logging.getLogger(__name__)
+
+# Ordered list of 2-tuples of (field, display_name) used to translate object
+# field names oto their human-readable form when calling .print_list().
+DISPLAY_FIELDS = (
+    ('id', 'ID'),
+    ('name', 'Name'),
+    ('description', 'Description'),
+    ('required_attributes', 'Required Attributes'),
+)
+
+# Fields to display when viewing a single record.
+VERBOSE_FIELDS = (
+    ('id', 'ID'),
+    ('name', 'Name'),
+    ('description', 'Description'),
+    ('required_attributes', 'Required Attributes'),
+    ('site', 'Site ID'),
+)
+
+
+# Main group
+@click.group()
+@click.pass_context
+def cli(ctx):
+    """
+    Protocol Type objects.
+
+    An Protocol Type resource can represent a network protocol type (e.g. bgp, is-is, ospf, etc.)
+
+    Protocol Types can have any number of required attributes as defined below.
+    """
+
+
+# Add
+@cli.command()
+@click.option(
+    '-r',
+    '--required_attributes',
+    metavar='REQUIRED ATTRIBUTES',
+    help='A key/value pair attached to this Protocol Type (format: key=value).',
+    multiple=True,
+    callback=callbacks.transform_attributes,
+    # required=True,
+)
+@click.option(
+    '-e',
+    '--description',
+    metavar='DESCRIPTION',
+    type=str,
+    help='The description for this Protocol Type.',
+)
+@click.option(
+    '-n',
+    '--name',
+    metavar='NAME',
+    type=str,
+    help='The name of the Protocol Type.',
+    required=True,
+)
+@click.option(
+    '-I',
+    '--id',
+    metavar='ID',
+    help='Unique ID of the Protocol Type being retrieved.',
+)
+@click.option(
+    '-s',
+    '--site-id',
+    metavar='SITE_ID',
+    type=int,
+    help='Unique ID of the Site this Protocol Type is under.  [required]',
+    callback=callbacks.process_site_id,
+)
+@click.pass_context
+def add(ctx, required_attributes, description, id, name, site_id):
+    """
+    Add a new Protocol Type.
+
+    You must provide a Protocol Type name or ID using the UPDATE option.
+
+    When adding a new Protocol Type, you must provide a value for the -n/--name
+    option.
+
+    Examples: OSPF, BGP, etc.
+
+    You must also provide attributes, you may specify the -a/--attributes
```

Comment:
Since regular `key=value` attributes aren't relevant to Protocol Types, this paragraph should be replaced with how `-r/--required_attributes` should be used. "You *may* also provide required Protocol attributes... They must exist first...", etc.